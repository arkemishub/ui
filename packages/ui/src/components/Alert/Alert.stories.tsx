/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { IAlertProps } from "./Alert.types";
import { Alert } from "../Alert";
import React, { useState } from "react";

const variants: IAlertProps[] = [
  { color: "error", title: "Error" },
  { color: "warning", title: "Warning" },
  { color: "info", title: "Info" },
  { color: "success", title: "Success" },
];

export default {
  title: "Alert",
  component: Alert,
};

export const Default = (args: Partial<IAlertProps>) => {
  return (
    <div className="flex flex-col gap-4">
      <Alert {...args}>
        <p className="text-sm ml-7">Here alert message</p>
      </Alert>
      {variants.map((item, index) => (
        <Alert key={index} {...args} color={item.color} title={item.title}>
          <p className="text-sm ml-7">Here alert message</p>
        </Alert>
      ))}
    </div>
  );
};
