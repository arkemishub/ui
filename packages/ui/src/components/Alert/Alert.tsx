/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

"use client";

import React, { MouseEventHandler, useMemo, useState } from "react";
import { twMerge } from "tailwind-merge";
import { Button } from "../Button";
import { IAlertProps } from "./Alert.types";

const variants: IAlertProps[] = [
  {
    color: "error",
    icon: (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        fill="none"
        viewBox="0 0 24 24"
        strokeWidth={1.5}
        stroke="currentColor"
        className="h-6 w-5"
      >
        <path
          strokeLinecap="round"
          strokeLinejoin="round"
          d="M12 9v3.75m9-.75a9 9 0 11-18 0 9 9 0 0118 0zm-9 3.75h.008v.008H12v-.008z"
        />
      </svg>
    ),
  },
  {
    color: "warning",
    icon: (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        fill="none"
        viewBox="0 0 24 24"
        strokeWidth={1.5}
        stroke="currentColor"
        className="h-6 w-5"
      >
        <path
          strokeLinecap="round"
          strokeLinejoin="round"
          d="M12 9v3.75m-9.303 3.376c-.866 1.5.217 3.374 1.948 3.374h14.71c1.73 0 2.813-1.874 1.948-3.374L13.949 3.378c-.866-1.5-3.032-1.5-3.898 0L2.697 16.126zM12 15.75h.007v.008H12v-.008z"
        />
      </svg>
    ),
  },
  {
    color: "info",
    icon: (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        fill="none"
        viewBox="0 0 24 24"
        strokeWidth={1.5}
        stroke="currentColor"
        className="h-6 w-5"
      >
        <path
          strokeLinecap="round"
          strokeLinejoin="round"
          d="M11.25 11.25l.041-.02a.75.75 0 011.063.852l-.708 2.836a.75.75 0 001.063.853l.041-.021M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9-3.75h.008v.008H12V8.25z"
        />
      </svg>
    ),
  },
  {
    color: "success",
    icon: (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        fill="none"
        viewBox="0 0 24 24"
        strokeWidth={1.5}
        stroke="currentColor"
        className="h-6 w-5"
      >
        <path
          strokeLinecap="round"
          strokeLinejoin="round"
          d="M9 12.75L11.25 15 15 9.75M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
        />
      </svg>
    ),
  },
];

const DeleteButton = ({
  onClick,
}: {
  onClick: MouseEventHandler<HTMLButtonElement>;
}) => (
  <Button onClick={onClick}>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      fill="none"
      viewBox="0 0 24 24"
      strokeWidth={1.5}
      stroke="currentColor"
      className="h-5 w-5"
    >
      <path
        strokeLinecap="round"
        strokeLinejoin="round"
        d="M6 18L18 6M6 6l12 12"
      />
    </svg>
  </Button>
);

function Alert({
  children,
  color,
  title,
  className,
  onDismiss,
  dismissible = false,
  icon,
}: IAlertProps) {
  const [visible, setVisible] = useState(true);
  const selected = useMemo(
    () => variants.find((variant) => variant.color == color),
    [color]
  );
  return (
    <>
      {visible && (
        <span
          role="presentation"
          className={twMerge(
            "alert",
            color === "error" && "alert--error",
            color === "warning" && "alert--warning",
            color === "info" && "alert--info",
            color === "success" && "alert--success",
            className
          )}
        >
          <div className="flex justify-between">
            <div className="flex flex-col">
              <div className="flex gap-2">
                {icon ?? selected?.icon ?? null}
                {title}
              </div>
              {children}
            </div>
            {dismissible && (
              <div className="align-center flex">
                <DeleteButton
                  onClick={() => {
                    setVisible(false);
                    onDismiss?.();
                  }}
                />
              </div>
            )}
          </div>
        </span>
      )}
    </>
  );
}

export default Alert;
