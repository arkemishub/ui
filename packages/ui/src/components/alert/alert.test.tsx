/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { act, render, screen } from "@testing-library/react";
import Alert from "./alert";

describe("Alert", () => {
  test("should match snapshot", () => {
    const { asFragment } = render(<Alert>Alert</Alert>);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render Alert component", () => {
    render(<Alert>Alert</Alert>);
    expect(screen.getByText("Alert")).toBeInTheDocument();
  });

  test("should add custom class", () => {
    render(<Alert className="custom-class">Alert</Alert>);
    expect(screen.getByRole("presentation")).toHaveClass("custom-class");
  });

  test("should render dismiss button when dismissible", () => {
    render(<Alert dismissible>Alert</Alert>);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test("should call onDismiss prop when dismiss button is clicked", () => {
    const onDismiss = jest.fn();
    render(
      <Alert dismissible onDismiss={onDismiss}>
        Alert
      </Alert>
    );
    const button = screen.getByRole("button");
    act(() => {
      button.click();
    });

    expect(onDismiss).toHaveBeenCalledTimes(1);
  });

  it("should render error alert", () => {
    render(<Alert color="error">Alert</Alert>);
    expect(screen.getByRole("presentation")).toHaveClass("bg-error");
  });

  it("should render warning alert", () => {
    render(<Alert color="warning">Alert</Alert>);
    expect(screen.getByRole("presentation")).toHaveClass("bg-warning");
  });

  it("should render info alert", () => {
    render(<Alert color="info">Alert</Alert>);
    expect(screen.getByRole("presentation")).toHaveClass("bg-info");
  });

  it("should render success alert", () => {
    render(<Alert color="success">Alert</Alert>);
    expect(screen.getByRole("presentation")).toHaveClass("bg-success");
  });
});
