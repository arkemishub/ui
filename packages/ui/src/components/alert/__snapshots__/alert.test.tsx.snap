// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-4 alert"
    role="presentation"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="flex flex-col"
      >
        <div
          class="flex gap-2"
        />
        Alert
      </div>
    </div>
  </div>
</DocumentFragment>
`;
